# 2. Hafta Lab-2: Regularization
### BGM 565: Siber Güvenlik için Makine Öğrenme Yöntemleri
### İstanbul Şehir Üni. - Bilgi Güvenliği Müh.
#### Dr. Ferhat Özgür Çatak

Bu lab çalışmasında doğrusal regresyon probleminin *gradient descent* çözümünde kullanılan **regularization** yöntemini sentetik bir kümesine uygulayacağız.

In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import SGDRegressor

Pandas kütüphanesi kullanarak ds2.txt dosyası verikumesi degişkenine (dataframe) atanmaktadır.

In [2]:
# veri kumesini oku
verikumesi = pd.read_csv("ds2.txt",delimiter="\t")
X = verikumesi.iloc[:,:-1].values
y = verikumesi.iloc[:,X.shape[1]].values

$X$ ve $y$ kullanılarak regresyon modeli l1 için eğitilir.

In [3]:
# modeli l1 regularization ile tanimla
regressor = SGDRegressor(penalty="l1", verbose=0, max_iter=100000)
# modeli egit
regressor.fit(X, y)
print(regressor.intercept_, regressor.coef_)

[0.14807012] [1.92528755 2.99112228 0.97849403 1.95101789]


Bu veri kümesi içi hipotez: $h(\mathbf{x}) = 0.1537206 + 1.92059957 \cdot x_1 + 2.98959435 \cdot x_2 + 0.97867187 \cdot x_3 + 1.94799199 \cdot x_4$ (eski model $h(\mathbf{x})=0.12490622+1.9516536 \cdot x_1 + 2.98882317 \cdot x_2 + 0.97638019 \cdot x_3 + 1.96358802 \cdot x4$). Hipotez kullanılarak $X$ verikümesinin çıktı değerleri tahmin ederek y_pred vektörüne atayalım

In [4]:
y_pred_l1 = regressor.predict(X)

$X$ ve $y$ kullanılarak regresyon modeli l2 için eğitilir.

In [5]:
# modeli l2 regularization ile tanimla
regressor = SGDRegressor(penalty="l2", verbose=0, max_iter=100000)
# modeli egit
regressor.fit(X, y)
print(regressor.intercept_, regressor.coef_)
y_pred_l2 = regressor.predict(X)

[0.15366447] [1.92069278 2.9895874  0.9786463  1.94803955]


Analitik : $h(\mathbf{x})=0.12490622+1.9516536 \cdot x_1 + 2.98882317 \cdot x_2 + 0.97638019 \cdot x_3 + 1.96358802 \cdot x_4$

L1-reg.  : $h(\mathbf{x}) = 0.1537206 + 1.92059957 \cdot x_1 + 2.98959435 \cdot x_2 + 0.97867187 \cdot x_3 + 1.94799199 \cdot x_4$

L2-reg.  : $h(\mathbf{x}) = 0.15366447 + 1.92069278 \cdot x_1 + 2.9895874 \cdot x_2 + 0.97864630 \cdot x_3 + 1.94803955 \cdot x_4$

In [6]:
df = pd.DataFrame({"y":y,"y_pred_l1":y_pred_l1,"y_pred_l2":y_pred_l2})
df

y  y_pred_l1  y_pred_l2
0  3.74   3.737806   3.739072
1  4.55   4.531350   4.531290
2  5.24   5.244872   5.244047
3  5.92   5.943300   5.942185
4  5.75   5.740648   5.739543
5  6.43   6.401441   6.400426
6  3.91   3.941119   3.941335
7  5.31   5.333337   5.333688
8  3.89   3.858173   3.858533
9  3.02   3.029537   3.031417